### PR TITLE
Check log link HTTP status codes before reporting and work with Fileserver_Maintenance jobs

### PR DESCRIPTION
**Pull request by daymanc**
**Resolves Tickets:**
https://www.pivotaltracker.com/story/show/92061084
https://www.pivotaltracker.com/story/show/92077774

### CODE REVIEW

**Comment by novas0x2a** [comment]
It seems to have incorrectly reported http://smurf/builds/piston/cowabunga/debug/functional-test-3.7.15724-12270.log as 404; curl does not know how to decode gzip, so it doesn't set `Accept-Encoding: gzip`; so it gets a 404 on the link.

Only user-agents that can accept gzip-encoded files and decompress them will be able to see the log, but that doesn't mean it's not there.

(https://www.pivotaltracker.com/n/projects/802327/stories/92405860)

File: file-bug.sh [hunk 1]
```diff
@@ -1,34 +1,62 @@
...
 
-if [ -z "$FILESERVER" -o -z "$RELEASE_VERSION" -o -z "$GITHUB_OWNER" -o -z "$TEST_BUILD_NUMBER" -o -z "$PIVOTAL_TOKEN" -o -z "$PIVOTAL_PROJECT_ID" -o -z "$PIVOTAL_OWNER_ID" ]
+if [ -z "$FILESERVER" -o -z "$RELEASE_VERSION" -o -z "$GITHUB_OWNER" -o -z "$TEST_BUILD_NUMBER" \
+     -o -z "$PIVOTAL_TOKEN" -o -z "$PIVOTAL_PROJECT_ID" -o -z "$PIVOTAL_OWNER_ID" -o -z $BUILD_NUMBER ]
 then
     echo "Required variable not set" >&2
     exit 1
 fi
 
-if [ "$GITHUB_OWNER" != "piston" ]; then
```
Comment (novas0x2a) [inline file-bug.sh]: the formatting of this code is [already correct](https://google-styleguide.googlecode.com/svn/trunk/shell.xml?showone=Loops#Loops), please don't change it.

File: file-bug.sh [hunk 2]
```diff
@@ -1,34 +1,62 @@
...
+    # the whole system
+    case "$(curl -o /dev/null --silent  --write-out '%{http_code}\n' $LINK)" in
+        *"200"*)                                          
+            DESCRIPTION="$DESCRIPTION\n$LINK"
+            ;;                                                             
+        *"404"*)                                      
+            DESCRIPTION="$DESCRIPTION\n[404/NotFound] => $LINK\nDid the tests actually run??"
+            ;;                                                             
+        *)                                             
+            DESCRIPTION="Error: The log links didn't provide a return code we could parse!\n\Build system is currently unstable in it's log reporting."
```
Comment (novas0x2a) [inline file-bug.sh]: its

File: file-bug.sh [hunk 3]
```diff
@@ -1,34 +1,62 @@
...
+                  "http://$FILESERVER/builds/$GITHUB_OWNER/$GITHUB_BRANCH/debug/functional-test-$RELEASE_VERSION-$TEST_BUILD_NUMBER.log.gz")
+
+DESCRIPTION="$FILEBUG_LINK\n$FUNCTIONAL_LINK\n"
+
+for LINK in "${FILESERVER_LINKS[@]}"
+do
+    # If we get anything that's not a 200 or a 404 there's something really wrong, file a bug on
+    # the whole system
+    case "$(curl -o /dev/null --silent  --write-out '%{http_code}\n' $LINK)" in
+        *"200"*)
```
Comment (novas0x2a) [inline file-bug.sh]: why the double *? it should just be a 200\n, right?

File: file-bug.sh (hunk 1 above)
Comment (daymanc) [inline file-bug.sh]: this change was strictly for code consistency with the other conditionals, i'll revert it

File: file-bug.sh (hunk 3 above)
Comment (daymanc) [inline file-bug.sh]: indeed i was going to take those out thanks for the catch

File: file-bug.sh (hunk 2 above)
Comment (daymanc) [inline file-bug.sh]: check

**Comment by daymanc** [comment]
@novas0x2a adding the 'accept-encoding' header with curl looks to resolve it:

$ curl -H 'Accept-encoding: gzip'  -o /dev/null --silent  --write-out '%{http_code}\n' http://smurf/builds/piston/cowabunga/debug/functional-test-3.7.15724-12270.log 
200           <=== happy

$ curl   -o /dev/null --silent  --write-out '%{http_code}\n' http://smurf/builds/piston/cowabunga/debug/functional-test-3.7.15724-12270.log 
404           <=== ! happy

File: file-bug.sh [hunk 4]
```diff
@@ -1,34 +1,61 @@
...
     exit 1
 fi
 
 if [ "$GITHUB_OWNER" != "piston" ]; then
+    echo "Repo owner is not piston, no bug will be filed."
     exit 0
 fi
 
-if ! [ "$TEST_GITHUB_OWNER" = "piston" -o -z "$TEST_GITHUB_OWNER" ]; then
+if ! [ "$TEST_GITHUB_OWNER" = "piston" -o -z "$TEST_GITHUB_OWNER" ]
```
Comment (novas0x2a) [inline file-bug.sh]: https://google-styleguide.googlecode.com/svn/trunk/shell.xml?showone=Loops#Loops

File: file-bug.sh [hunk 5]
```diff
@@ -1,34 +1,61 @@
...
-
+LABEL_ID="5803519"
+FILEBUG_LINK="Bug: https://albino.piston.cc/job/X_Report_Failed_Test/$BUILD_NUMBER/console"
+FUNCTIONAL_LINK="Tests: https://albino.piston.cc/job/Functional_Tests/$TEST_BUILD_NUMBER/console"
+FILESERVER_LINKS=("http://$FILESERVER/builds/$GITHUB_OWNER/$GITHUB_BRANCH/debug/functional-test-$RELEASE_VERSION-$TEST_BUILD_NUMBER.log"
+                  "http://$FILESERVER/builds/$GITHUB_OWNER/$GITHUB_BRANCH/debug/functional-test-$RELEASE_VERSION-$TEST_BUILD_NUMBER.log.gz")
+
+DESCRIPTION="$FILEBUG_LINK\n$FUNCTIONAL_LINK\n"
+
+for LINK in "${FILESERVER_LINKS[@]}"
```
Comment (novas0x2a) [inline file-bug.sh]: https://google-styleguide.googlecode.com/svn/trunk/shell.xml?showone=Loops#Loops

**Comment by novas0x2a** [comment]
is there an example of a mis-filed bug that this is supposed to catch?

**Comment by daymanc** [comment]
@novas0x2a would have caught: https://www.pivotaltracker.com/n/projects/802327/stories/92276820

File: file-bug.sh (hunk 4 above)
Comment (daymanc) [inline file-bug.sh]: script was inconsistent from the start, but i'll make the changes to keep the 'if/then' inline as  suggested as well as with the loop. to be safe going forward i'll assume the google style guide is also the piston style guide in these regards

**Comment by novas0x2a** [comment]
This would have annotated the link with a 404 message, sure, but it wouldn't have reduced the effort required for the sluthing or anything; it just saves @fcforrest the 5 seconds it takes to click the link.

A more direct fix might be to set a variable like "LOGS_EXPECTED" or something and only print the log links if it's set, which couldn't be foolproof without more integration with run-functional-tests (maybe you could grep the output for "Placing start marker in syslog" or something, but that's pretty fragile and liable to cause more problems than it fixes).

I mean, the simplest fix which would be just as effective as all of these other options is to make file-bug unconditionally add this message: 
> If the functional test logs are missing from smurf, the two most likely causes are 1) the test failed to start correctly (check the albino log) or 2) this is an old bug and the logs have been deleted.

**Comment by daymanc** [comment]
My original change and comment was regarding just pulling from source (run-functionals) which is telling the rest of the world anyway. If run-functionals didn't run/generate logs, it would basically be the same as the proposed LOGS_EXPECTED var (empty). If pulling from run-functionals (via a marker/delimiter) is considered fragile then one would assume run-functionals output is considered fragile to some degree as well. Just looking at the options we have hanging around out there I still stand behind that change hehe :) If we want to dump everything and just add that string to get this moving forward cool, just want to get traction on this and move forward.

**Comment by daymanc** [comment]
@fcforrest any comments you'd like to add?

**Comment by fcforrest** [comment]
I believe that this PR addresses the initial problems that I was observing.

1. If the log links fail to resolve, at least it's obvious.
2. The numbers for the pentos build and functional run seem to line up 100% of the time, and in the event they don't, I now have a link to the Failed_Test job that created them, to diagnose.

As for additional improvements to the bug-reporting interface, I have myriad suggestions, but they can be captured in follow-on tickets.

**Comment by daymanc** [comment]
@fcforrest agreed there are a good number of things that can be added to this script. Hopefully the changes address what is currently required as well as the FS maintenance integration. Nice profile photo BTW..

**Comment by daymanc** [comment]
@fcforrest looks like we are on the same page for phase 1, can we go ahead and get this merged? Thanks.

**Pull request merged**